### PR TITLE
bpo-32604: PEP 554 for use in test suite

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
[Issue #22](https://github.com/ericsnowcurrently/multi-core-python/issues/22) is about making use of subinterpreters (specifically the PEP 554 API) in the CPython test suite.
Until PEP 554 is accepted (or if it isn't) we should add a PEP 554 implementation as Lib/test/support/_interpreters.py (e.g. test.support._interpreters).

https://bugs.python.org/issue39881

### CODE REVIEW

**Comment by nanjekyejoannah** [comment]
I intentionally left out isolated as it looks like its still experimental.

**Comment by nanjekyejoannah** [comment]
ohh waiting on send and recv is non trivial I guess like in https://github.com/python/cpython/pull/19829/files ,  guess.

File: Doc/library/_interpreters.rst
Comment (ericsnowcurrently) [inline Doc/library/_interpreters.rst]: We shouldn't add the module to the stdlib documentation yet.  However, it might still be worth adding this file in Lib/test/support as a reference.

File: Doc/library/index.rst
Comment (ericsnowcurrently) [inline Doc/library/index.rst]: This should wait until we add the module to the stdlib.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: The file should be `interpreters.py` instead of `_interpreters.py`.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: There are some missing:

```suggestion
__all__ = ['Interpreter', 'SendChannel', 'RecvChannel',
            'is_shareable', 'create_channel',
            'list_all_channels', 'get_current',
            'get_main', 'create', 'list_all', 'RunFailedError']
```

The channel-related exception types should be added too.

FWIW, I sometimes find it helpful to group the related names:

```suggestion
__all__ = [
    'Interpreter', 'get_current', 'get_main', 'create', 'list_all',
    'SendChannel', 'RecvChannel',
    'create_channel', 'list_all_channels', 'is_shareable',
    'RunFailedError',
    ]
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]:
```suggestion
    id = _interpreters.create(isolated=isolated)
```

Note that the flag on `_xxsubinterpreters.create()` doesn't exist yet.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]:
```suggestion
    return Interpreter(id, isolated=isolated)
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: Sorry, I forgot to add this to the PEP:

```suggestion

    @property
    def isolated(self):
        if self._isolated is None:
            self._isolated = _interpreters.is_isolated(self._id)
        return self._isolated

```

Note that `_xxsubinterpreters.is_isolated()` does not exist yet.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]:
```suggestion
    def __init__(self, id, *, isolated=None):
        self._id = id
        self._isolated = isolated
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: This should have a docstring.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]:
```suggestion
        This blocks the current Python thread until done.
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: I'm guessing this was for debugging.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: For the docstrings in the Python module you don't need to document the signature.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: This will never be raised from `recv()`.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: This will never be raised from `recv_nowait()`.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: This will never be raised from `send()`.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: This will never be raised from `send_nowait()`.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: I'm guessing that this (and the similar code in the other methods) was for debugging.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: The ID should be read-only.

```suggestion

    def __init__(self, id):
        self._id = id

    @property
    def id(self):
        return self._id
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]:
```suggestion
    def recv(self, *, _delay=10 / 1000):  # seconds
        """ channel_recv() -> obj

        Get the next object from the channel,
        and wait if none have been sent.
        Associate the interpreter with the channel.
        """
        sentinel = object()
        obj = _interpreters.channel_recv(self._id, sentinel)
        while obj is sentinel:
            time.sleep(_delay)
            obj = _interpreters.channel_recv(self._id, sentinel)
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]:
```suggestion
    _NOT_SET object()
    
    def recv_nowait(self, default=_NOT_SET):
        """recv_nowait(default=None) -> object

        Like recv(), but return the default
        instead of waiting.
        """
        if default is _NOT_SET:
            return _interpreters.channel_recv(self.id)
        else:
            return _interpreters.channel_recv(self.id, default)
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: docstring

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: The ID should be read-only.

```suggestion
    def __init__(self, id):
        self._id = id
        
    @property
    def id(self):
        return self._id
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: This probably isn't needed.  It's a bit cleaner to in-line where needed since there are only a couple of places that use this.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: These should come from `_xxsubinterpreters`.

```suggestion
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: WIth #19829 this will be the following:

```suggestion
        wait = _interpreters.channel_send_wait(self.id, obj)
        wait()
```

Until then you have no way of knowing when the object has been received.  Sleeping is not correct, but is fine temporarily.

```suggestion
        _interpreters.channel_send(self.id, obj)
        time.sleep(2)
```

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/_interpreters.py]: The recv doesn't really fit.  Until #19829 lands you can't return anything so you might as well always return `False`:

```suggestion
        _interpreters.channel_send(self.id, obj)
        return False
```

Once #19829 lands this code will look like this:

```suggestion
        return _interpreters.channel_send(self.id, obj)
```

or:

```suggestion
        wait = _interpreters.channel_send_wait(self.id, obj)
        return wait(timeout=0)
```

File: Modules/_xxsubinterpretersmodule.c
Comment (ericsnowcurrently) [inline Modules/_xxsubinterpretersmodule.c]: What are these for?

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for working on this, @nanjekyejoannah!  I left a number of comments for you.  Hopefully it helps.  I'd be glad to chat about any of this if you have questions.

FYI, I haven't looked closely at the tests yet.

**Comment by vstinner** [comment]
I marked bpo-39881 as a duplicate of bpo-32604. I changed this PR title to use bpo-32604.

File: Lib/test/support/_interpreters.py
Comment (nanjekyejoannah) [reply Lib/test/support/_interpreters.py]: yes

File: Lib/test/support/_interpreters.py
Comment (nanjekyejoannah) [reply Lib/test/support/_interpreters.py]: Making this readonly makes it hard to return RecvCnannel in methods like `create_channel()`. what do you think?

File: Lib/test/support/_interpreters.py
Comment (nanjekyejoannah) [reply Lib/test/support/_interpreters.py]: Same comment as above. Making this readonly makes it hard to return SendChannel in methods like `create_channel()` . what do you think?

File: Lib/test/support/_interpreters.py
Comment (nanjekyejoannah) [reply Lib/test/support/_interpreters.py]: I wonder then what the high-level implementation of these exceptions looks like if we delete this. Any ideas?

File: Modules/_xxsubinterpretersmodule.c
Comment (nanjekyejoannah) [reply Modules/_xxsubinterpretersmodule.c]: These Exceptions I just added their lower level support.

**Review by nanjekyejoannah — COMMENTED** [review]
@ericsnowcurrently, I made the requested changes and I think I left some notes on making ``id`` fields read-only for SendChannel and RecvChannel.  And about the high-level implementation of Exceptions. well, we can merge a minimal version and I can follow up with more PRs after feature freeze.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [reply Lib/test/support/_interpreters.py]: We should just expose the exceptions from the low-level module.

File: Lib/test/support/_interpreters.py
Comment (ericsnowcurrently) [reply Lib/test/support/_interpreters.py]: They're only read-only to users.  Regardless, I'm not clear on why we would change the value ever, even in create_channel().

File: Modules/_xxsubinterpretersmodule.c
Comment (ericsnowcurrently) [reply Modules/_xxsubinterpretersmodule.c]: We don't use them anywhere.  They aren't in the PEP anymore either.

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:18]:
```suggestion
    'ChannelError', 'ChannelNotFoundError',
    'ChannelEmptyError', 'ChannelNotEmptyError', 'NotReceivedError',
    ]
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py]: Just expose the low-level function.

```suggestion
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:3]: The simplest approach with some of the pieces is to expose them directly from the low-level module:

```suggestion
import _xxsubinterpreters as _interpreters
# aliases:
from _xxsubinterpreters import (
    ChannelError, ChannelNotFoundError,
    ChannelEmptyError, ChannelNotEmptyError, NotReceivedError,
    is_shareable,
)
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py]:
```suggestion
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py]:
```suggestion
        _interpreters.run_string(self._id, src_str)
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py]:
```suggestion
    def recv(self, *, _delay=10 / 1000):  # 10 milliseconds
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py]: It would be better to use a private sentinel instead of None.  Otherwise it will be confusing for users when they pass None as the default.  `_NOT_SET needs to move to module-level due to the weird scoping rules of class bodies.

```suggestion
    def recv_nowait(self, default=_NOT_SET):
        """
        Like recv(), but return the default
        instead of waiting.
        """

        if default is _NOT_SET:
            return _interpreters.channel_recv(self._id)
        else:
            return _interpreters.channel_recv(self._id, default)

_NOT_SET = object()
```

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for taking the time to make those changes, @nanjekyejoannah.  I've left a few more comments, but it mostly looks good.  Like you said, we can add onto this after it is merged.

One other thing I recommend doing in this PR is leaving a comment at each place where the implementation is blocked on low-level implementation (e.g. `channel_send_wait()`).

File: Lib/test/support/interpreters.py
Comment (nanjekyejoannah) [reply Lib/test/support/interpreters.py]: It looks like we can not directly use `_NOT_SET` as a default argument for a this kind of method in a class I think. Unless we do something like this:
```
def recv_nowait(self, default=None):
        """
        Like recv(), but return the default
        instead of waiting.
        """
        if default is None:
            default = _NOT_SET
        if default is _NOT_SET:
            return _interpreters.channel_recv(self._id)
        else:
            return _interpreters.channel_recv(self._id, default)
```

**Comment by nanjekyejoannah** [comment]
It looks like we can not directly use `_NOT_SET` as a default argument for this kind of method in a class I think. Unless we do something like this:
```
def recv_nowait(self, default=None):
        """
        Like recv(), but return the default
        instead of waiting.
        """
        if default is None:
            default = _NOT_SET
        if default is _NOT_SET:
            return _interpreters.channel_recv(self._id)
        else:
            return _interpreters.channel_recv(self._id, default)
```

**Comment by ericsnowcurrently** [comment]
@nanjekyejoannah, sorry I wasn't more clear.  I meant `_NOT_SET` should be moved to module-level scope.  It would be something like this:

```python
_NOT_SET = object()


class RecvChannel:
    ...

    def recv_nowait(self, default=_NOT_SET):
        """
        Like recv(), but return the default
        instead of waiting.
        """
        if default is _NOT_SET:
            return _interpreters.channel_recv(self._id)
        else:
            return _interpreters.channel_recv(self._id, default)
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:27]: formatting (PEP 8):

```suggestion


```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:35]: formatting (PEP 8):

```suggestion


```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:43]: formatting (PEP 8):

```suggestion


```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:105]: formatting (PEP 8):

```suggestion


```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:153]: formatting (PEP 8):

```suggestion


```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:154]: formatting (PEP 8):

```suggestion

```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py]: Move this above `class RecvChannel`.  I suppose you *could* move it to the top of the file, but it usually makes sense to keep things near where they're actually needed.  I'll leave a comment showing where I'd put it.

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py]: Once you move `_NOT_SET` above `class RecvChannel` in the module, this should work:

```suggestion
    def recv_nowait(self, default=_NOT_SET):
        """
        Like recv(), but return the default
        instead of waiting.

        This function is blocked by a missing low-level
        implementation of channel_recv_wait().
        """
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:118]: Here's where I'd put `_NOT_SET`:

```suggestion
_NOT_SET = object()


class RecvChannel:
```

File: Lib/test/support/interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/support/interpreters.py:10]: formatting (PEP 8):

```suggestion


```

File: Lib/test/test_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/test_interpreters.py:10]: formatting (PEP 8):

```suggestion


```

File: Lib/test/test_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/test_interpreters.py:22]: formatting (PEP 8):

```suggestion


```

File: Lib/test/test_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/test_interpreters.py:32]: formatting (PEP 8):

```suggestion


```

File: Lib/test/test_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/test_interpreters.py:39]: formatting (PEP 8):

```suggestion


```

File: Lib/test/test_interpreters.py
Comment (ericsnowcurrently) [inline Lib/test/test_interpreters.py:525]: formatting (PEP 8):

```suggestion
```

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for making more changes.  Sorry for adding extra work.  It looks like there's only a handful of things left to do and we can merge this:

- [x] sort out `_NOT_SET` for recv_nowait()`
- [x] add comments for the pieces that are not implemented yet in the low-level module
- [x] remove the changes (2 added exceptions) in `Modules/_xxsubinterpretersmodule.c`

Other than that, I've left a bunch of comments for formatting (PEP 8) changes (though they don't relate to correctness).

I considered recommending that we move the testing helpers to `test.support`, but we can worry about that later. :slightly_smiling_face:

I also considered recommending that we limit the tests on the new `interpreters` module to just verifying integration with the low-level module.  However, there is merit to having functional tests here.  So what you have looks good.

**Comment by nanjekyejoannah** [comment]
I have made the requested changes; please review again

**Review by ericsnowcurrently — APPROVED** [review]
Looks good.  There are a few things to adjust, particularly once we finish the low-level module.  However, what you have here gives us nearly everything else.  Thanks for working on this!

Feel free to merge this after Łukasz lets everyone know that beta1 is done (and the 3.9 branch created).  I expect that will be tomorrow.

**Comment by nanjekyejoannah** [comment]
> 
> 
> Looks good. There are a few things to adjust, particularly once we finish the low-level module. However, what you have here gives us nearly everything else. Thanks for working on this!
> 
> Feel free to merge this after Łukasz lets everyone know that beta1 is done (and the 3.9 branch created). I expect that will be tomorrow.

Thanks for the review @ericsnowcurrently . I will merge this tommorrow.

**Comment by pablogsal** [comment]
Seems that `test_interpreters` is leaking on master:

https://buildbot.python.org/all/#/builders/394/builds/94

......
test_interpreters leaked [216, 216, 216] references, sum=648
test_interpreters leaked [84, 84, 84] memory blocks, sum=252
1 test failed again:
    test_interpreters

**Comment by pablogsal** [comment]
Opened https://bugs.python.org/issue40783

**Pull request merged**
